### PR TITLE
Agent build order

### PR DESCRIPTION
**Pull request by nickwalton**

### CODE REVIEW

File: holodeck/environments.py [hunk 1]
```diff
@@ -112,14 +115,14 @@ def info(self):
...
-        for agent in self._all_agents:
+        for agent in self.agents:
             result.append("\tName: ")
             result.append(agent.name)
             result.append("\n\tType: ")
             result.append(type(agent).__name__)
             result.append("\n\t")
             result.append("Sensors:\n")
-            for sensor in self._sensor_map[agent.name].keys():
+            for sensor in agent.sensors.items():
```
Comment (MaxRobinsonTheGreat) [inline holodeck/environments.py]: Items returns the key value pair, not the values. Should be getting values

File: holodeck/environments.py [hunk 2]
```diff
@@ -263,26 +268,26 @@ def set_state(self, agent_name, location, rotation, velocity, angular_velocity):
...
-            location (np.ndarray or list): The position to spawn the agent in the world, in XYZ coordinates (in meters).
-        """
-        self._add_agents(agent_definition)
-        if self._agent is None:
-            self._agent = self.agents[agent_definition.name]
-        self._enqueue_command(SpawnAgentCommand(location, agent_definition.name, agent_definition.type.agent_type))
+    def add_agent(self, agent_def):
+
+        if agent_def.name in self.agents:
+            print("Error: agent name duplicate.")
```
Comment (MaxRobinsonTheGreat) [inline holodeck/environments.py]: Throw error, don't print

File: holodeck/tests/agent_pytest.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+    env.add_agent(agent)
+    state = env.tick()
+    env.set_state("uav1", loc, rot, (0, 0, 0), (0, 0, 0))
+    state = env.tick()
+
+    sensed_loc = state["uav1"]["LocationSensor"]
+    sensed_rot = state["uav1"]["RotationSensor"]
+    assert almost_equal(loc, sensed_loc)
+    assert almost_equal(rot, sensed_rot)
```
Comment (MaxRobinsonTheGreat) [inline holodeck/tests/agent_pytest.py]: The greatest offense to mankind. No newline at the end of the file. For shame

File: holodeck/environments.py [hunk 4]
```diff
@@ -93,6 +92,10 @@ def __init__(self, agent_definitions=None, binary_path=None, scenario_key=None,
         
         # Flag indicates if the user has called .reset() before .tick() and .step()
         self._initial_reset = False
+        self.reset()
+
+        for agent_def in agent_definitions:
+            self.add_agent(agent_def)
```
Comment (kolbytn) [inline holodeck/environments.py]: duplicate from reset. remove.

File: holodeck/environments.py [hunk 5]
```diff
@@ -171,8 +175,10 @@ def reset(self):
         # Load agents
         self.agents = dict()
         self._state_dict = dict()
-        self._load_existing_agents(self._initial_agents)
         self.load_scenario()
+        for agent_def in self._initial_agents:
+            self.add_agent(agent_def)
```
Comment (kolbytn) [inline holodeck/environments.py]: Add existing agents before loading the scenario.

File: holodeck/environments.py [hunk 6]
```diff
@@ -146,7 +147,10 @@ def load_scenario(self):
                                                 location=sensor['location'],
                                                 rotation=sensor['rotation'],
                                                 params=params))
-            self.agents[agent['agent_name']].add_new_sensors(sensors)
+
+            agent_def = AgentDefinition(agent['agent_name'], agent['agent_type'], starting_loc = agent["location"], sensors=sensors)
```
Comment (kolbytn) [inline holodeck/environments.py]: remove spaces after starting_loc

**Pull request merged**